crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2850,21 +2850,31 @@ impl<'db> Type<'db> {
                         self,
                         [
                             Signature::new(
-                                Parameters::new([Parameter::positional_only(Some(
-                                    Name::new_static("o"),
-                                ))
-                                .with_annotated_type(Type::any())
+                                Parameters::new([Parameter::positional_or_keyword(
+                                    Name::new_static("object"),
+                                )
+                                .with_annotated_type(Type::object(db))
                                 .with_default_type(Type::string_literal(db, ""))]),
                                 Some(KnownClass::Str.to_instance(db)),
                             ),
                             Signature::new(
                                 Parameters::new([
-                                    Parameter::positional_only(Some(Name::new_static("o")))
-                                        .with_annotated_type(Type::any()), // TODO: ReadableBuffer
-                                    Parameter::positional_only(Some(Name::new_static("encoding")))
-                                        .with_annotated_type(KnownClass::Str.to_instance(db)),
-                                    Parameter::positional_only(Some(Name::new_static("errors")))
-                                        .with_annotated_type(KnownClass::Str.to_instance(db)),
+                                    Parameter::positional_or_keyword(Name::new_static("object"))
+                                        // TODO: Should be `ReadableBuffer` instead of this union type:
+                                        .with_annotated_type(UnionType::from_elements(
+                                            db,
+                                            [
+                                                KnownClass::Bytes.to_instance(db),
+                                                KnownClass::Bytearray.to_instance(db),
+                                            ],
+                                        ))
+                                        .with_default_type(Type::bytes_literal(db, b"")),
+                                    Parameter::positional_or_keyword(Name::new_static("encoding"))
+                                        .with_annotated_type(KnownClass::Str.to_instance(db))
+                                        .with_default_type(Type::string_literal(db, "utf-8")),
+                                    Parameter::positional_or_keyword(Name::new_static("errors"))
+                                        .with_annotated_type(KnownClass::Str.to_instance(db))
+                                        .with_default_type(Type::string_literal(db, "strict")),
```

Comment:
interesting: these overloads differ from typeshed's overloads in that `str(encoding='utf8')` is permitted by these overloads, but not by typeshed's. It's hard to say whether it should or shouldn't be permitted, really: it works at runtime, but arguably the type checker would be doing the user a favour by telling the user off for doing it, since there's no reason why it would ever be useful (it's almost certainly indicative of a user error)

Reply:
Oh, right. I didn't realize that `object` is required in the second overload in typeshed. Should we just ignore the discrepancy? Change it in Red Knot? Change it in typeshed?

I agree that it's useless to do `str(encoding='utf-8')`, but it is valid at runtime, clearly specified as such in the [documentation](https://docs.python.org/3/library/stdtypes.html#str), and seemingly harmless. I also doubt that we would catch actual errors by disallowing it. So I'm inclined to say that it should be changed in typeshed?

Reply:
hmm, that's a fair point about the documentation. I still weakly prefer the typeshed stubs as they are: I just can't think of any reason why a user would want to write `str(encoding='utf8')` when they could just as well write `str()` or `""`. I think the only reason why it would ever appear in a codebase would be as a result of a typo or a bad codemod, so I'd prefer for type checkers to complain about it.

I really don't have a strong opinion, though! If you feel like sending a PR to typeshed, feel free -- the other typeshed maintainers might feel differently 😄

Reply:
I don't feel strongly about it, and not really familiar with typeshed's policies when it comes to questions like this, so I guess it's fine to just leave it as a discrepancy.

Reply:
I think typeshed should be changed: this is a classic case of "what should a type checker report" vs "what should a linter report". Perfectly reasonable to write a lint rule saying "this call is silly", but IMO not reasonable for a type checker to claim the call is a type error, when it clearly is not.

Reply:
(But "thinking typeshed should be changed" is as far as I will take this, I don't care enough to submit a PR and follow through on debating it 😆 )

Reply:
> Perfectly reasonable to write a lint rule saying "this call is silly", but IMO not reasonable for a type checker to claim the call is a type error, when it clearly is not.

hmm, sounds like there could be a gap in the market for a type-aware linter here 🤔